[fx][normalize] Allow for args to be left as args

jfix71 opened the pull request:
Summary: Normalization is kind of broken currently. But making default arguments visible still appears to work, and is nice functionality to still be able to rely on/use. Adds an option to `NormalizeArgs`'s `__init__` called `normalize_to_only_use_kwargs` which defaults to true, which if set to false will keep using the same signature as provided, but additionally set kwargs in kwargs.

Test Plan: Added test to `test_fx_experimental`.

Differential Revision: D27759448

ansley commented on torch/fx/experimental/normalize.py:
```diff
@@ -80,13 +83,13 @@ def call_module(self, target : Target, args : Tuple[Argument, ...], kwargs : Dic
-                new_kwargs = self._args_kwargs_to_normalized_kwargs(sig, args, kwargs)
-                if new_kwargs:
-                    return super().call_module(target, (), new_kwargs)
+                new_args, new_kwargs = self._args_kwargs_to_normalized_args_kwargs(sig, args, kwargs)
```
Will tuple unpacking fail if you return `None`?

jamesr66a commented on torch/fx/experimental/normalize.py:
```diff
@@ -80,13 +83,13 @@ def call_module(self, target : Target, args : Tuple[Argument, ...], kwargs : Dic
-    def _args_kwargs_to_normalized_kwargs(self, sig : inspect.Signature, args : Tuple[Argument, ...],
-                                          kwargs : Dict[str, Any]) -> Optional[Dict[str, Any]]:
+    def _args_kwargs_to_normalized_args_kwargs(self, sig : inspect.Signature, args : Tuple[Argument, ...],
+                                          kwargs : Dict[str, Any]) -> Optional[Tuple[Tuple[Any], Dict[str, Any]]]:
```
This return type is getting pretty complex. Can we define a class or NamedTuple for the return?

jamesr66a commented on torch/fx/experimental/normalize.py:
```diff
@@ -27,13 +28,15 @@ class NormalizeArgs(Transformer):
         super().__init__(module)
         self.normalize_functionals = normalize_functionals
         self.normalize_modules = normalize_modules
+        self.normalize_to_only_use_kwargs = normalize_to_only_use_kwargs
```
I think this will still be broken for operators. Can you add a flag `self.normalize_operators` that defaults to False here?

jfix71 replied:
I'm a little confused on this one -- AFAICT `NormalizeArgs` and `NormalizeOperators` are totally independent. What would `self.normalize_operators` do inside `NormalizeArgs`?

jfix71 replied:
Friendly ping @jamesr66a 🙂

jamesr66a replied:
The `else` branch starting at line 65 normalizes arguments for nodes with operator targets. `NormalizeOperators` changes operator targets, not args

jfix71 replied:
Got it, added.

jfix71 commented:
Reimplemented on top of https://github.com/pytorch/pytorch/pull/55992

jamesr66a reviewed (approved):
Looks good, thanks!

jfix71 committed:
[fx][normalize] Allow for args to be left as args (#55995)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/55995

Normalization is kind of broken currently. But making default arguments visible still appears to work, and is nice functionality to still be able to rely on/use. Adds an option to `NormalizeArgs`'s `__init__` called `normalize_to_only_use_kwargs` which defaults to true, which if set to false will keep using the same signature as provided, but additionally set kwargs in kwargs.

Test Plan: Added test to `test_fx_experimental`.

Reviewed By: 842974287

Differential Revision: D27759448

fbshipit-source-id: 21c972b18a7335e433b28d6c2452e410c1f0fa21
